### PR TITLE
Many0: overhaul, additional test, dedicated error

### PR DESCRIPTION
**Pull request by tstorch**
Rewrite of the `many0` macro (shorter and more comprehensible), now with own ErrorKind, and more tests for this macro.

One `assert_eq` will fail in current master due to wrong `tag` macro in the tests in `macros.rs`. Once #184 is merged this test will also succeed.

### CODE REVIEW

**Comment by Geal** [comment]
Hi! I'm finally taking the time to review that pull request, sorry for the delay. The reason I jumped through hoops with `Vec` creation is that reallocations can be a bit costly. I'll take your version anyhow, since it's much cleaner, and see what I can improve from there.

File: src/util.rs [hunk 1]
```diff
@@ -641,28 +642,29 @@ pub fn error_to_u32<E>(e: &ErrorKind<E>) -> u32 {
...
-    ErrorKind::LengthValueFn             => 22,
-    ErrorKind::Eof                       => 23,
-    ErrorKind::ExprOpt                   => 24,
-    ErrorKind::ExprRes                   => 25,
-    ErrorKind::CondReduce                => 26,
-    ErrorKind::Switch                    => 27,
-    ErrorKind::TagBits                   => 28,
-    ErrorKind::OneOf                     => 29,
-    ErrorKind::NoneOf                    => 30,
+    ErrorKind::Many0                     => 9,
```
Comment (Geal) [inline src/util.rs]: error codes should not be reordered

**Comment by tstorch** [comment]
No Problem. Nice to hear that it's in :smile: 

What should I do about the error codes?

**Comment by Geal** [comment]
Nothing, I'm taking care of it while merging the PR :)

**Comment by tstorch** [comment]
That's how I like to work :grin:

**Pull request merged**